plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -748,27 +800,32 @@ fu_genesys_usbhub_device_setup(FuDevice *device, GError **error)
 				   (guint8 *)&self->static_ts,
 				   sizeof(FuGenesysStaticToolString));
 	}
-	if (self->static_ts.tool_string_version != 0xff) {
-		gchar rev[3] = {0};
+	if (memcmp(&self->static_ts.tool_string_version, "USB", 3) == 0) {
```

Comment:
`if (memcmp(&self->static_ts, "USB", 3) == 0) {`makes more sense to me as `tool_string_version` is one byte long.

OOI:
- Also, does the `USB` string is a prefix or the full string?
- Does the string `USB`... is reported by the firmware itself or by some USB stack (linux)?
- I wonder if the check should be done in the function `fu_genesys_usbhub_device_get_descriptor_data()` so it would be available for other Tool String data. What do you think?

Reply:
My opinion is it makes sense to have two checks "USB" first, and tool string version then `0xff`

Reply:
> - Does the string USB... is reported by the firmware itself or by some USB stack (linux)?

`USB String Descriptor` reported by the firmware.

---
> - Also, does the USB string is a prefix or the full string? 

It is a prefix, returned values are like
- `09:33:17:0613 FuPluginGenesys      Static info:    55 53 42 32 2e 30 20 48 75 62 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00`, the string is "USB2.0 Hub"
- `09:33:17:0358 FuPluginGenesys      Static info:    55 53 42 33 2e 30 20 48 75 62 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00`, the string is "USB3.0 Hub";
---
> - I wonder if the check should be done in the function fu_genesys_usbhub_device_get_descriptor_data() so it would be available for other Tool String data. What do you think?

A hub unsupported tool string still can update firmware by this plugin, but we don't need to do that at this phase :). So check the `static_ts` at first is fine.

Reply:
> My opinion is it makes sense to have two checks "USB" first, and tool string version then 0xff

I'm sorry to mislead the `tool_string_version` value in `0xff`, as I said in last comment, the tool string was replied by firmware. In-house usage alters the `tool_string_version` to `0xff` to indicate unsupported of tool string, but we don't alter here, therefore I remove the `0xff` checking.

Reply:
> if (memcmp(&self->static_ts, "USB", 3) == 0) {makes more sense to me as tool_string_version is one byte long.

OK, let me modify that.

Reply:
> I'm sorry to mislead the tool_string_version value in 0xff, as I said in last comment, the tool string was replied by firmware. In-house usage alters the tool_string_version to 0xff to indicate unsupported of tool string, but we don't alter here, therefore I remove the 0xff checking.

Well, to make sure I get it right; none of the firmware (current and past) would reply a static tool string version value of `0xff`? which means the previous check was dead code, right?

And you replaced the check by `"USB"` because you found `3520` does not reply a static tool string structure, right?

Just for curiosity, I found two GL3521-22 (with genesys VID/PID) in my devices with firmware `92.??` (do not recall the version). Can I found firmware upgrade for these devices? Also, is that possible to buy GL3520 and GL3590 end-products somewhere?

Reply:
>Well, to make sure I get it right; none of the firmware (current and past) would reply a static tool string version value of 0xff? which means the previous check was dead code, right?

Yes, the `0xff` check is copied from windows code. For linux code base I gave to you, is supporting for GL3523 and GL3590 only. But for other product, there has extra check code.

---
> And you replaced the check by "USB" because you found 3520 does not reply a static tool string structure, right?

Yes, USB string descriptor replied by firmware. And our USB hub replies "Product String Descriptor" for unknown index of String Descriptor. BTW, "Product String Descriptor" default is "USB2.0 Hub" or "USB2.1 Hub" for our USB3.x hub(pure 2.0 hub does not support).

---
>Just for curiosity, I found two GL3521-22 (with genesys VID/PID) in my devices with firmware 92.?? (do not recall the version). Can I found firmware upgrade for these devices?

You can do these,
1. Dump firmware from the flash.
2. Erase flash to run mask code. 
3. Update the dumped firmware.

---
> Also, is that possible to buy GL3520 and GL3590 end-products somewhere?

I don't know how many products are there, you can ask for our PMs and sales to query more info.